includes/class-wp-document-revisions-validate-structure.php

```diff
@@ -0,0 +1,628 @@
+		}
+
+		add_action( 'admin_menu', array( &$this, 'add_menu' ), 20 );
+
+		add_action( 'admin_enqueue_scripts', array( &$this, 'enqueue_scripts' ) );
+		add_action( 'rest_api_init', array( &$this, 'wpdr_register_route' ) );
+	}
+
+	/**
+	 * Provides support to call functions of the parent class natively.
+	 *
+	 * @since 3.4.0
+	 *
+	 * @param function $function the function to call.
+	 * @param array    $args the arguments to pass to the function.
+	 * @returns mixed the result of the function.
+	 */
+	public function __call( $function, $args ) {
+		return call_user_func_array( array( &self::$parent, $function ), $args );
+	}
+
+	/**
+	 * Add settings menu page.
+	 *
+	 * @since 3.4.0
+	 **/
+	public static function add_menu() {
+		$slug = 'wpdr_validate';
+		add_submenu_page( 'edit.php?post_type=document', __( 'Validate Structure', 'wp_document_revisions' ), __( 'Validate Structure', 'wp_document_revisions' ), 'edit_documents', $slug, array( __CLASS__, 'page_validate' ) );
+
+		// help text.
+		add_action( 'load-document_page_' . $slug, array( __CLASS__, 'add_help_tab' ) );
+	}
+
+	/**
+	 * Register route
+	 */
+	public function wpdr_register_route() {
+		$args = array(
+			'methods'             => 'PUT',
+			'callback'            => array( &$this, 'correct_document' ),
+			'permission_callback' => array( &$this, 'check_permission' ),
+			'args'                => array(
+				'id'   => array(
+					'required'          => true,
+					'sanitize_callback' => 'absint',
+				),
+				'code' => array(
+					'required'          => true,
+					'sanitize_callback' => 'absint',
+				),
+				'parm' => array(
+					'required'          => true,
+					'sanitize_callback' => 'absint',
+				),
+			),
+		);
+
+		register_rest_route(
+			'wpdr/v1',
+			'correct/(?P<id>[\d]+)/type/(?P<code>[\d]+)/attach(?P<parm>[\d]+)',
+			$args
+		);
+	}
+
+	/**
+	 * Rest function to correct document structure data.
+	 *
+	 * @since 3.4.0
+	 *
+	 * @param WP_REST_Request $request the arguments to pass to the function.
+	 * @return WP_REST_Response
+	 */
+	public function correct_document( $request ) {
+		$params = $request->get_params();
+		$id     = $params['id'];
+		$parm   = $params['parm'];
+		if ( 4 === $params['code'] ) {
+			// Attachment exists but post_content does not contain it.
+			// revalidate input values.
+			$content = get_post_field( 'post_content', $id, 'db' );
+			if ( false !== self::$parent->extract_document_id( $content ) || get_post_field( 'post_parent', $parm, 'db' ) !== $id ) {
+				return new WP_Error( 'inconsistent_parms', __( 'Inconsistent data sent to Interface', 'wp-document-revisions' ) );
+			}
+			if ( empty( $content ) || false === strpos( $content, '<' ) ) {
+				$content = (string) $parm;
+			} else {
+				$content = self::$parent->format_doc_id( $parm );
+			}
+			global $wpdb;
+			// phpcs:disable WordPress.DB.PreparedSQL.InterpolatedNotPrepared,WordPress.DB.PreparedSQL.NotPrepared,WordPress.DB.DirectDatabaseQuery
```

Comment:
It might be good to include a comment indicating why this needs to be a direct DB query.

Reply:
Two parts to this.
I started using WordPress after I retired. I worked for many years on a commercial ERP with some 2000 DB tables. SQL holds no fear for me - and I do find the continuing warnings here somewhat wearing.

However in this specific module, it is deliberate.
a) At root, my objective is to validate the document structure - which are held in the database. So I go to the database.
b) I want only 4 fields from the document - so I get just them. Yes,, I could do a WP_Query for that - but why?
c) Yes, I did hesitate over the level of access. If I gave it to admin, then clearly I would expect the `edit_post` capability would not filter out many. However, I reflected that whilst there is an amount of fixing that can be done, there may be many more times that the resolution will be to reload the document. And who should have that - those that can edit it, not administrators. Yes, this may mean that many records returned get filtered out.
d) Finally to fix the data, for metadata I use the standard functions. But for the Post, I use a direct update. Why? Because to use the WP functions would create a revision  that would create an invalid revision. This makes no sense.

I believe that a future enhancement could be to have a validate option on the "All Documents" screen. The issue for me is the management of the feedback row.

I will put some design notes at the top of the module.